### ISSUE TITLE
IDP Initiated Authentication Script

### ISSUE THREAD

**Issue by nynymike** [GluuFederation/oxAuth#267]
![image](https://cloud.githubusercontent.com/assets/3717101/17334198/f827312e-589a-11e6-8a61-aa9918ba7809.png)

It would be nice if oxAuth could act as a gateway for consolidating IDP-initiated authentication. The basic idea is that the Gluu Server would be the SP--i.e. the authentication is coming from an external organization's IDP. Gluu would consume the assertion, dynamically register the person if necessary, and then redirect the person as specified by the relay_state parameter. 

It's unclear if this can be done within an interception script alone, or if we'd need to implement a new endpoint. Also, would the oxAuth publish SP metadata for this endpoint? This service should generate its own keys and certificates.

For testing, use a [Shibboleth IDP](https://wiki.shibboleth.net/confluence/display/SHIB2/IdPUnsolicitedSSO), which support IDP initiated as the "internal IDP" 

Sequence Diagram Source

```
Title IDP Initiated Authentication

Browser->Website 1: GET Website 1
Website 1->Browser: Redirect to internal IDP
Browser->Internal IDP: Login
Internal IDP->Browser:
Browser->Internal IDP: Request IDP initiated authn with relay_state=gluu_server_url
Browser->Gluu Server: Post SAML Assertion to Gluu Server SAML SP ACS
Gluu Server->Gluu Server: Validate SAML Assertion
Gluu Server->Gluu Server: Dynamic Enrollment if user doesn't exist
Gluu Server->Gluu Server: Create Gluu Server Session
Gluu Server->Gluu Server: Map Source IDP to Website 2
Gluu Server->Browser: Redirect to website 2 relay_state URL
Browser->Website 2: GET website 2 relay_state URL
Website 2->Gluu Server: Get user claims
```

**Comment by yurem** [comment]
There are 3 oxAuth questions regarding this diagram:
1) In all our existing workflows we create session on authorization request. As result we can't just send SAML Request to existing endpoints. We need new endpoint which will create session and start authentication workflow.
2) How oxAuth should know which acr_values should use?
3) In last step there process to get user claims. Which authorization flow should we use? Where and when we start it?

Also I think we have to communicate with Internal SP according to this [diagram](https://en.wikipedia.org/wiki/SAML_2.0#SP_Redirect_Artifact.3B_IdP_Redirect_Artifact). These are steps 8 and 9.

**Comment by nynymike** [comment]
1.  I agree... I was saying the same.
 2.  I think the acr would be static. There can only be one script to process IDP iniitated authn.
 3.  The IDP will send user claims in the assertion, which are used for dynamic enrollment.  Post-enrollment, the interception script should redirect to the url speciifed in the relay_state. This application should be a normal OpenID Connect RP, and won't know anything about the IDP intitiated flow. It will send for authorization, but hopefully the user will already have a session in the browser.

**Comment by arvindsinghtomar** [comment]
@nynymike @yurem can you please explain what we are trying to achieve here?

**Comment by arvindsinghtomar** [comment]
I talked to will regarding this issue.
As far I understand requirements we need to achieve similar functionality like the inbound-saml  passport .I think instead of creating new script again we can use the passport for this.

**Comment by yurem** [comment]
We can implement next flow which uses Gluu-Passport:
![inbound saml](https://user-images.githubusercontent.com/6104200/39752768-19d3190c-52c5-11e8-878f-e3338733b03f.png)
```
title Inbound SAML

User->Browser: Click on page
Browser->IDP: Authenticate User
IDP->Passport: SAML2 POST AuthN
Passport->Passport: Validate request, check signature, etc...
Passport->oxAuth: AuthZ request with custom parameter(s) with user profile
oxAuth->Script: Invoce custom script "authenticate" method
Script->oxAuth: Validate userprofile and authenticated user by uid/e-mail
oxAuth->App: Send access_token
```
Here si plan:
1. Add new enpoint to trigger Inbound SAML flow,
2. RequestingPartyId -> OpenID client mapping. We need new json config file:
```
relayingPartyId_1: {oxAuth client details needed for AuthZ},
relayingPartyId_2: {oxAuth client details needed for AuthZ},
relayingPartyId_3: {oxAuth client details needed for AuthZ},
```
3. Prepare AuthZ request and send user profile as signed JWT via custom paramter ([exmaple](https://github.com/GluuFederation/oxTrust/blob/master/saml-openid-auth-client/src/main/java/org/gluu/oxauth/client/authentication/AuthenticationFilter.java#L183)),
4. Write new custom script which trust user Profile (check JWT signature) which Passport send in AuthZ request,
5. Store Passport configuration in LDAP and add oxTrust GUI to change it.

We can finish all steps except 5 in 3.1.4ю

**Comment by jgomer2001** [comment]
I'm closing this, I think the recently added IDP-initiated capabilities to passport do the job.

**Comment by nynymike** [comment]
Are there docs on how to configure it in Gluu? I end up having to address this question several times a week.

**Comment by jgomer2001** [comment]
Not yet I think since it was recently introduced. Actually, it was solved for 3.1.5. It will be part of 3.1.5 docs
